model/riscv_stateen_regs.sail

```diff
@@ -0,0 +1,484 @@
+
+function clause currentlyEnabled(Ext_Smstateen) = hartSupports(Ext_Smstateen)
+function clause currentlyEnabled(Ext_Ssstateen) = hartSupports(Ext_Ssstateen)
+
+function enable_mstateen() -> bool = hartSupports(Ext_Smstateen)
+function enable_sstateen() -> bool = currentlyEnabled(Ext_S) & (currentlyEnabled(Ext_Smstateen) | currentlyEnabled(Ext_Ssstateen))
+function enable_hstateen() -> bool = false & (currentlyEnabled(Ext_Smstateen) | currentlyEnabled(Ext_Ssstateen))
+
+bitfield Mstateen0 : bits(64) = {
+  SE0     : 63,
+  ENVCFG  : 62,
+
+  CSRIND  : 60,
+  AIA     : 59,
+  IMSIC   : 58,
+  CONTEXT : 57,
+  P1P13   : 56,
+  P1P14   : 55,
+  CTR     : 54,
+
+  JVT     : 2,
+  FCSR    : 1,
+  C       : 0,
+}
+
+// TODO: This needs to be modified once support Hypervisor is added
+// see https://github.com/riscv/sail-riscv/pull/910#discussion_r2070685435
+function enable_SE0() -> bool = currentlyEnabled(Ext_S) | misa[F] == 0b0
+
+function legalize_mstateen0(m : Mstateen0, v : bits(64)) -> Mstateen0 = {
+  let v = Mk_Mstateen0(v);
+
+  [m with
+    SE0     = if enable_SE0() then v[SE0] else 0b0,
+    ENVCFG  = if currentlyEnabled(Ext_S) then v[ENVCFG] else 0b0,
+
+    CSRIND  = 0b0, // Not implemented yet
+    AIA     = 0b0, // Not implemented yet
+    IMSIC   = 0b0, // Not implemented yet
+    CONTEXT = 0b0, // Not implemented yet
+    P1P13   = 0b0, // Not implemented yet
+    P1P14   = 0b0, // Not implemented yet
+    CTR     = 0b0, // Not implemented yet
+
+    JVT     = 0b0, // Not implemented yet
+    FCSR    = if misa[F] == 0b1 then 0b0 else v[FCSR],
+    C       = 0b0,
+  ];
+}
+
+register mstateen0 : Mstateen0 = legalize_mstateen0(Mk_Mstateen0(zeros()), zeros())
+
+mapping clause csr_name_map = 0x30C  <-> "mstateen0"
+mapping clause csr_name_map = 0x31C  <-> "mstateen0h"
+
+function clause is_CSR_defined(0x30C) = enable_mstateen()
+function clause is_CSR_defined(0x31C) = enable_mstateen() & (xlen == 32)
+
+function clause read_CSR(0x30C) = mstateen0.bits[xlen - 1 .. 0]
+function clause read_CSR(0x31C if xlen == 32) = mstateen0.bits[63 .. 32]
+
+function clause write_CSR((0x30C, value) if xlen == 32) = { mstateen0 = legalize_mstateen0(mstateen0, mstateen0.bits[63 .. 32] @ value); mstateen0.bits[31 .. 0] }
+function clause write_CSR((0x30C, value) if xlen == 64) = { mstateen0 = legalize_mstateen0(mstateen0, value); mstateen0.bits }
+function clause write_CSR((0x31C, value) if xlen == 32) = { mstateen0 = legalize_mstateen0(mstateen0, value @ mstateen0.bits[31 .. 0]); mstateen0.bits[63 .. 32] }
+
+bitfield Mstateen1 : bits(64) = {
+  SE1 : 63, // There is no official name for bit 63 yet
+}
+
+function legalize_mstateen1(m : Mstateen1, v : bits(64)) -> Mstateen1 = {
+  let v = Mk_Mstateen1(v);
+
+  [m with
+    SE1 = 0b0
+  ];
+}
+
+register mstateen1 : Mstateen1 = legalize_mstateen1(Mk_Mstateen1(zeros()), zeros())
+
+mapping clause csr_name_map = 0x30D  <-> "mstateen1"
```

Comment:
It's minor but I think it would be a lot easier to read if the 0, 1, 2, 3 were grouped together for all the lines. Going to be way easier to see copy/paste errors that way too.